<a href="https://colab.research.google.com/github/Catheria29/HomePricesPrediction/blob/main/HomePricesPrediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import kagglehub
import pandas as pd
import os #is used to import Operating system, is a built-in Python library that allows your code to interact with the operating system your script is running on (whether that's Linux in Google Colab, Windows on your laptop, or macOS)
path = kagglehub.dataset_download("dansbecker/melbourne-housing-snapshot")#is a folder
print("Files in folder:", os.listdir(path))#list all files in the folder
file_path = os.path.join(path, 'melb_data.csv')#simple operation to get the correct path to the file
melbourne_data = pd.read_csv(file_path)#load the data as a dataframe
print(melbourne_data.head())#prints the first five rows and the title of each column
melbourne_data.describe()# gives statical description of the data
#count: The number of non-empty rows in that column.
#mean: The mathematical average.Formula: mean=(∑x)/n​ (The sum of all values divided by the number of values).This tells you the "typical" price or size, but it can be skewed by extremely expensive mansions.
#std (Standard Deviation): .This measures how "spread out" the numbers are. High std: The prices vary wildly (some houses are $400k, others are $5 million).Low std: The values are all very close to the average
#min: The smallest value found in that column.The Insight: In this dataset, it shows you the cheapest house sold or the smallest plot of land.
#max: The largest value found in that column.The Insight: This identifies the most expensive house in the entire dataset.
#75% (Third Quartile):75% of the houses sold for less than this value (or, the top 25% of houses sold for more than this). The Insight: This represents the "premium" end of the market.
#50% (Median): The middle value when all rows are sorted from lowest to highest. Note: If the mean is much higher than the 50% median, it means a few super-expensive "outlier" houses are pulling the average up.
#25% (First Quartile):25% of the houses sold for less than this value. The Insight: If you are a budget buyer, this represents the "entry-level" price point of the market.

Using Colab cache for faster access to the 'melbourne-housing-snapshot' dataset.
Files in folder: ['melb_data.csv']
       Suburb           Address  ...             Regionname Propertycount
0  Abbotsford      85 Turner St  ...  Northern Metropolitan        4019.0
1  Abbotsford   25 Bloomburg St  ...  Northern Metropolitan        4019.0
2  Abbotsford      5 Charles St  ...  Northern Metropolitan        4019.0
3  Abbotsford  40 Federation La  ...  Northern Metropolitan        4019.0
4  Abbotsford       55a Park St  ...  Northern Metropolitan        4019.0

[5 rows x 21 columns]


Rooms         Price  ...    Longtitude  Propertycount
count  13580.000000  1.358000e+04  ...  13580.000000   13580.000000
mean       2.937997  1.075684e+06  ...    144.995216    7454.417378
std        0.955748  6.393107e+05  ...      0.103916    4378.581772
min        1.000000  8.500000e+04  ...    144.431810     249.000000
25%        2.000000  6.500000e+05  ...    144.929600    4380.000000
50%        3.000000  9.030000e+05  ...    145.000100    6555.000000
75%        3.000000  1.330000e+06  ...    145.058305   10331.000000
max       10.000000  9.000000e+06  ...    145.526350   21650.000000

[8 rows x 13 columns]

In [2]:
melbourne_data.columns#title or  name of each column


Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude',
       'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

In [4]:
melbourne_data = melbourne_data.dropna(axis=0)#check each row, if there is a value missing in any column, it will remove the hole row

In [3]:
y = melbourne_data.Price#the tatget, value to predict
melbourne_features = ['Rooms', 'Bathroom', 'Landsize', 'Lattitude', 'Longtitude']#list of the choseen features from the original dataset
X = melbourne_data[melbourne_features]#retrieving the column whose name are in the list melbourne_features data from the original dataset
X.describe()

Rooms      Bathroom       Landsize     Lattitude    Longtitude
count  13580.000000  13580.000000   13580.000000  13580.000000  13580.000000
mean       2.937997      1.534242     558.416127    -37.809203    144.995216
std        0.955748      0.691712    3990.669241      0.079260      0.103916
min        1.000000      0.000000       0.000000    -38.182550    144.431810
25%        2.000000      1.000000     177.000000    -37.856822    144.929600
50%        3.000000      1.000000     440.000000    -37.802355    145.000100
75%        3.000000      2.000000     651.000000    -37.756400    145.058305
max       10.000000      8.000000  433014.000000    -37.408530    145.526350

In [6]:
X.head()

Rooms  Bathroom  Landsize  Lattitude  Longtitude
1      2       1.0     156.0   -37.8079    144.9934
2      3       2.0     134.0   -37.8093    144.9944
4      4       1.0     120.0   -37.8072    144.9941
6      3       2.0     245.0   -37.8024    144.9993
7      2       1.0     256.0   -37.8060    144.9954

In [7]:
# from sklearn.tree import DecisionTreeRegressor#used to create models. Scikit-learn is easily the most popular library for modeling the types of data typically stored in DataFrames.
# melbourne_model = DecisionTreeRegressor(random_state=1)#regression = numbers, random_state = consistency
# melbourne_model.fit(X, y)#feed the features and known targets, to Capture patterns from provided data. This is the heart of modeling.
# """The steps to building and using a model are:

#     Define: What type of model will it be? A decision tree? Some other type of model? Some other parameters of the model type are specified too.
#     Fit: Capture patterns from provided data. This is the heart of modeling.
#     Predict: Just what it sounds like
#     Evaluate: Determine how accurate the model's predictions are.
# """

DecisionTreeRegressor(random_state=1)

In [8]:
# print("Making predictions for the following 5 houses:")
# print(X.head())
# print("The predictions are")
# print(melbourne_model.predict(X.head()))#testing the model

Making predictions for the following 5 houses:
   Rooms  Bathroom  Landsize  Lattitude  Longtitude
1      2       1.0     156.0   -37.8079    144.9934
2      3       2.0     134.0   -37.8093    144.9944
4      4       1.0     120.0   -37.8072    144.9941
6      3       2.0     245.0   -37.8024    144.9993
7      2       1.0     256.0   -37.8060    144.9954
The predictions are
[1035000. 1465000. 1600000. 1876000. 1636000.]


In this part, we are going to explore how to use data, to get a better model that works accurately (at least mostly) not only for the "in sample"(training data and validation data are the same) but also for the "out_of_sample data" (a portionof data was isolated from the training data using: train_X, val_X, train_y, val_y = train_test_split(X, y, random_state = 0), this part is called validation data. it will be used to test the model and calculate the mean absolute error

In [4]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error
from sklearn.tree import DecisionTreeRegressor

train_X, val_X, train_y, val_y = train_test_split(X,y,random_state=0)#returns numpu arrays
melbourne_model = DecisionTreeRegressor()
melbourne_model.fit(train_X,train_y)
val_prediction = melbourne_model.predict(val_X)
print(mean_absolute_error(val_y,val_prediction))
print(val_y.describe())




246418.66863033874
count    3.395000e+03
mean     1.076293e+06
std      6.509603e+05
min      2.100000e+05
25%      6.500000e+05
50%      8.950000e+05
75%      1.341750e+06
max      9.000000e+06
Name: Price, dtype: float64


From this result , MAE = 246418.66863033874 and the average of the house price in the validation data is 1.076293e+06, so the error is about quarter of the average home value.

The phenomenon called **overfitting** (The "Memorizer"), where a model matches the training data almost perfectly, but does poorly in validation and other new data. ***Example***: When a tree has too many leaves, it basically "memorizes" the specific houses it was trained on.
When a model fails to capture important distinctions and patterns in the data, so it performs poorly even in training data, that is called **underfitting** (The "Generalizer"). ***Example***: When a tree is too shallow (too few leaves), it groups houses together that don't actually belong together.

**Investigate and optimize the decicion tree** **model** **bold text**

In [5]:
def get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y):
  model = DecisionTreeRegressor(max_leaf_nodes=max_leaf_nodes, random_state = 0)
  model.fit(train_X,train_y)
  predicts_val = model.predict(val_X)
  mae = mean_absolute_error(val_y, predicts_val)
  return mae

In [11]:
for i in [490,500,550, 600, 5000]:
  my_mae = get_mae(i, train_X, val_X, train_y, val_y)
  print("Max leaf nodes = %d \t\t Mean Absolute Error = %d"%(i,my_mae))

Max leaf nodes = 490 		 Mean Absolute Error = 231716
Max leaf nodes = 500 		 Mean Absolute Error = 231301
Max leaf nodes = 550 		 Mean Absolute Error = 232348
Max leaf nodes = 600 		 Mean Absolute Error = 233292
Max leaf nodes = 5000 		 Mean Absolute Error = 248846


We can ddeduce that the minimum Max leaf nodes = 500 		 Mean Absolute Error = 231301, thus, this can be used to fix our model accuracy if decision tree is still the choosen model.

In [ ]:
final_model = DecisionTreeRegressor(max_leaf_nodes=500, random_state = 0)
final_model.fit(X,y)

We'll look at the random forest as an example of better model to improve the accuracy of our prediction.

The random forest uses many trees, and it makes a prediction by averaging the predictions of each component tree. It generally has much better predictive accuracy than a single decision tree and it works well with default parameters.



In [12]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error

forest_model = RandomForestRegressor(random_state=1)
forest_model.fit(train_X, train_y)
melb_preds = forest_model.predict(val_X)
print(mean_absolute_error(val_y,melb_preds))

180860.37877504269


In conclusion, 180860 < 231301
Thus the Forest performs better than Tree.